### DISCUSSION TITLE
No output voltage from pwm at pin STM32f411ce

### DISCUSSION THREAD

**Discussion by Dave-Poissant** [zephyrproject-rtos/zephyr#50675] [Q&A] [answered]
Hi zephyr developers, Hello!

I'm currently trying to reproduce the "blinky pwm" sample onto my stm32f411ce and I'm facing different problems that I would like to clear asap. I'm working with the platformio ide env with zephyr release 2.7.1.

At this moment, I have made a pwm device as follows in my overlay file:

```
/ {
    aliases {
            i2c1node = &i2c1;
            i2c3node = &i2c3;
            spi1node = &spi1;
            uartnode = &usart2;
            pwmled0 = &pwm_led_white;
    };

    pwmleds {
        compatible = "pwm-leds";
        pwm_led_white: external_pwm_led {
            pwms = < &pwm1 1 256 PWM_POLARITY_NORMAL >;
        };
    };
};

&i2c3 { 
    pinctrl-0 = <&i2c3_scl_pa8 &i2c3_sda_pb4>;
    pinctrl-names = "default";
    status = "okay"; 
    clock-frequency = <I2C_BITRATE_STANDARD>; 
};

&spi1 {
    pinctrl-names = "default";
    cs-gpios = <&gpioa 0x4 GPIO_ACTIVE_LOW>;
};

&usart2 {
    status = "okay";
    pinctrl-0 = < &usart2_tx_pa2 &usart2_rx_pa3 >;
    pinctrl-names = "default";
    current-speed = < 0x1c200 >;
};

&dma1 {
    status = "okay";
};

&dma2 {
    status = "okay";
};

&timers1 {
    status = "okay";
    pwm1: pwm {
        status = "okay";
        pinctrl-0 = <&tim1_ch1n_pb13>;
        pinctrl-names = "default";
    };
};
```

My prj.conf file goes as follows:
```
# nothing here
CONFIG_LOG=y
CONFIG_LOG_PRINTK=y

CONFIG_DMA=y
CONFIG_UART_STM32=y
CONFIG_SERIAL=y
CONFIG_UART_ASYNC_API=y

# enable pinmux
CONFIG_PINMUX=y
CONFIG_PWM=y

CONFIG_GPIO=y
```

Here is how I'm trying to make my LED turn on:
```
#include "utils/LoggingSWO.h"
#include <kernel.h>
#include <device.h>
#include <drivers/uart.h>
#include <drivers/pwm.h>
#include <kernel.h>
#include <drivers/gpio.h>
#include <devicetree.h>

#define PWM_LED0_NODE	DT_ALIAS(pwmled0)

#if DT_NODE_HAS_STATUS(PWM_LED0_NODE, okay)
#define PWM_CTLR	       DT_PWMS_CTLR(PWM_LED0_NODE)
#define PWM_CHANNEL       DT_PWMS_CHANNEL(PWM_LED0_NODE)
#define PWM_FLAGS	       DT_PWMS_FLAGS(PWM_LED0_NODE)
#else
#error "Unsupported board: pwmled0 devicetree alias is not defined"
#define PWM_CTLR	       DT_INVALID_NODE
#define PWM_CHANNEL       0
#define PWM_FLAGS	       0
#endif

void main(void)
{
	SWOPrintf("Starting \r\n");

	uint32_t max_period;
	uint32_t period;
	uint8_t dir = 0U;
	int ret = -1;

	const struct device *pwm_dev = DEVICE_DT_GET(PWM_CTLR);
	if (!pwm_dev)
	{
		SWOPrintf("Cannot find PWM device!\n\r");
		return;
	}
	else
	{
		SWOPrintf("PWM device find\n\r");
		SWOPrintf("%s\n\r", pwm_dev->name);
	}
	
	if (!device_is_ready(pwm_dev)) {
		SWOPrintf("Error: PWM device %s is not ready\n\r", pwm_dev->name);
		return;
	}

	while (1) {
		led_is_on = !led_is_on;
		gpio_pin_set(dev, LED0_PIN, (int)led_is_on);
		
		SWOPrintf("Main looooooop!\r\n");

		ret = pwm_pin_set_usec(pwm_dev, PWM_CHANNEL,
				       256 , 128, PWM_FLAGS);
		if (ret != 0) {
			SWOPrintf("Error %d: failed to set pulse width\n\r", ret);
			return;
		}
	
		k_msleep(25U);
	}
}
```

My led is plugged in pin PB13 and Gnd(23) (reference in next picture):

![image](https://user-images.githubusercontent.com/43071732/192384946-ab870f5e-9454-46eb-8c91-adb1ee82f3f5.png)

I receive no error from the ```pwm_pin_set_usec(...)``` call, but I get no output voltage at pin PB13 like I would expect. Am I missing something during initialization or somewhere else?

**Comment by Dave-Poissant** [comment, marked as answer]
I've also tryed to add the flag GPIO_ACTIVE_HIGH to the pwm flags, but with no success (did the same test with GPIO_ACTIVE_LOW).

**Reply by Coi-l** [reply]
@Dave-Poissant  Did you ever find an answer to this. I seem to be facing a very similar, if not the same, issue.
I'm trying to get the pwm working for the STM32f11ce on a blackpill board. Everything seems fine; I can get the device from the device tree and execute pwm_pin_set on it with no errors. But I am not getting any results on the pins.

Any insight you might have would be greatly appreciated.

Thanks in advance!

**Reply by Dave-Poissant** [reply]
Yes my bad I forgot to answer the issue!

<img width="1180" alt="image" src="https://user-images.githubusercontent.com/43071732/203674372-24f51083-e3f3-47a2-a7eb-fe6231ecbec3.png">

Thoses pins are inverted pwm and seemed to be using other hardware to generate their pwm. Knowing this, i just switched pin with the same code and it worked.

In the 3.1.0 release i think there is a fix for that matter.

I think this was the intial [issue](https://github.com/zephyrproject-rtos/zephyr/issues/21384) related to this fix. You can see in stm32CubeIDE that it has a different API for those pwm pins:

`HAL_TIM_PWM_Start` vs `HAL_TIM_PWMN_Start`

**Reply by Coi-l** [reply]
Aha, alright. Thank you!
I actually found my issue and it was not the same as yours. In zephyr 3.1.0 the blackpill dts is missing a prescaler setting for timer4 which is what it uses for the defined pins pb6 and pb7. 
So I just added this in my overlay file
&timers4 {
        st,prescaler = <10>;
};

Thank you for taking the time to answer. Hopefully both our solutions can be of help for someone.
